proposals/p3720.md

```diff
@@ -0,0 +1,853 @@
+used with Carbon's `.(`**`)` and `->(`**`)` operators.
+
+For example, this is how I would expect C++ code to call the above Carbon
+functions:
+
+```cpp
+struct C {
+  int F() const { return m + 1; }
+  int m;
+};
+
+int main() {
+  // pointer to data member `m` of class C
+  int C::* p = &C::m;
+  C c = {2};
+  assert(c.*p == 2);
+  assert(Carbon::GetField(c, p) == 2);
+  Carbon::SetField(&c, p, 4);
+  assert(c.m == 4);
+  // pointer to method `F` of class C
+  int (C::*q)() const = &C::F;
+  assert(Carbon::CallMethod(&c, q) == 5);
+}
+```
+
+### Member forwarding
+
+Consider a class that we want to act like it has the members of another type.
+For example, a type `Box(T)` that has a pointer to a `T` object allocated on the
+heap:
+
+```carbon
+class Box(T:! type) {
+  var ptr: T*;
+  // ???
+}
+```
+
+`Box(T)` should act like it has all the members of `T`:
+
+```carbon
+class String {
+  fn Search[self: Self](c: u8) -> i32;
+}
+var b: Box(String) = ...;
+var position: i32 = b.Search(32);
+```
+
+There are two ingredients to make this work:
+
+-   We need some way to make `b.Search` be equivalent to `b.(String.Search)` not
+    `b.(Box(String).Search)`.
+-   We need the act of binding a method of `String` to a `Box(String)` value
+    work by dereferencing the pointer `b.ptr`.
+
+For the first ingredient, we need a way to customize
+[simple member access](/docs/design/expressions/member_access.md) for the class
+`Box(T)`. Normally simple member access on a value like `b`
+[looks in the type of b](/docs/design/expressions/member_access.md#values).
+However, types can designate other entities to also look up in using the
+`extend` keyword (see [proposal #2760](/proposals/p2760.md#proposal)):
+
+-   `extend impl as I` adds lookup into the implementation of an interface `I`.
+-   `extend base: B` adds lookup into the base class `B`.
+-   `extend adapt C` adds lookup into the adapted class `C`.
+-   [mixins are also planning](/proposals/p2760.md#future-work-mixins) on using
+    an `extend` syntax
+
+The natural choice is to add another way of using `extend` with consistent
+lookup rules as the other uses of `extend`, with conflicts handled as determined
+by leads issues
+[#2355](https://github.com/carbon-language/carbon-lang/issues/2355) and
+[#2745](https://github.com/carbon-language/carbon-lang/issues/2745). I propose
+`extend api T`:
+
+```carbon
+class Box(T:! type) {
+  var : T*;
+  extend api T;
+}
+```
+
+This means that lookup into `Box(T)` also looks in the type `T` for members. In
+this way, `b.Search` will find `b.(String.Search)`.
+
+> **Note:** This is an alternative to defining an `alias` for each member of the
+> extended type. This avoids having to repeat them, which is both lengthy and
+> could get out of sync as the class evolves. The `extend api` approach works
+> even if, as in this example, we don't know the names of the members of the
+> type being extended.
```

Comment:
I'm a little cautious of this, because it means we don't know the API of a type until the first call is done. Is there a use-case for allowing `extend api` on an incomplete type?

Reply:
To try to make this a little more concrete, here's another example:

```
base class C {
  fn F();
}

class D {
  extend base: C;
  extend api B;
  fn G[self: Self]();
}

fn D.G[self: Self]() {
  // Does compilation fail here?
  self.F();
}
```

Specifically, although `F` has a definition on the base, it could also be defined on the api B. Since `extend api` is just aliasing things that don't have a conflict, and `B.F` wouldn't be a conflict (since `D` can shadow `base.F` right?), I think the name lookup here requires the incomplete type. But a compilation failure here might be a little less clear for developers.

Reply:
I've changed it to restrict to complete types.

Reply:
`extend api` is being moved to its own proposal.